libs/ml/examples/distributed_learning/main.cpp

```diff
@@ -0,0 +1,420 @@
+#include "math/matrix_operations.hpp"
+#include "math/tensor.hpp"
+#include "ml/core/graph.hpp"
+#include "ml/dataloaders/mnist_loaders/mnist_loader.hpp"
+#include "ml/layers/fully_connected.hpp"
+#include "ml/ops/activation.hpp"
+#include "ml/ops/loss_functions/cross_entropy_loss.hpp"
+
+#include <algorithm>
+#include <chrono>
+#include <ctime>
+#include <fstream>
+#include <iostream>
+#include <mutex>
+#include <queue>
+#include <string>
+#include <thread>
+#include <utility>
+#include <vector>
+
+#define NUMBER_OF_CLIENTS 10
+#define NUMBER_OF_PEERS 3
+#define NUMBER_OF_ITERATIONS 20
+#define BATCH_SIZE 128
+#define SYNCHRONIZATION_INTERVAL 3
+#define LEARNING_RATE .001f
+#define TEST_SET_RATIO 0.03f
+
+using namespace fetch::ml::ops;
+using namespace fetch::ml::layers;
+
+using DataType         = float;
+using TensorType       = fetch::math::Tensor<DataType>;
+using TensorVectorType = std::vector<TensorType>;
+using SizeType         = fetch::math::SizeType;
+
+enum class CoordinatorState
+{
+  RUN,
+  STOP,
+};
+
+class Coordinator
+{
+public:
+  CoordinatorState state = CoordinatorState::RUN;
+};
+
+class TrainingClient
+{
+public:
+  TrainingClient(std::string const &images, std::string const &labels, std::string const &id)
+    : dataloader_(images, labels)
+    , id_(id)
+  {
+    dataloader_.SetTestRatio(test_set_ratio_);
+    dataloader_.SetRandomMode(true);
+    g_.AddNode<PlaceHolder<TensorType>>("Input", {});
+    g_.AddNode<FullyConnected<TensorType>>("FC1", {"Input"}, 28u * 28u, 10u);
+    g_.AddNode<Relu<TensorType>>("Relu1", {"FC1"});
+    g_.AddNode<FullyConnected<TensorType>>("FC2", {"Relu1"}, 10u, 10u);
+    g_.AddNode<Relu<TensorType>>("Relu2", {"FC1"});
+    g_.AddNode<FullyConnected<TensorType>>("FC3", {"Relu2"}, 10u, 10u);
+    g_.AddNode<Softmax<TensorType>>("Softmax", {"FC3"});
+    g_.AddNode<PlaceHolder<TensorType>>("Label", {});
+    g_.AddNode<CrossEntropyLoss<TensorType>>("Error", {"Softmax", "Label"});
+
+    // Clear loss file
+    std::ofstream lossfile("losses_" + id_ + ".csv", std::ofstream::out | std::ofstream::trunc);
+    lossfile.close();
+  }
+
+  void SetCoordinator(std::shared_ptr<Coordinator> &coordinator_ptr)
+  {
+    coordinator_ptr_ = coordinator_ptr;
+  }
+
+  /**
+   * Main loop that runs in thread
+   */
+  void MainLoop()
+  {
+    std::ofstream lossfile("losses_" + id_ + ".csv", std::ofstream::out | std::ofstream::app);
+    DataType      loss;
+
+    while (coordinator_ptr_->state == CoordinatorState::RUN)
+    {
+      // Shuffle the peers list to get new contact for next update
+      fetch::random::Shuffle(gen_, peers_, peers_);
+
+      // Train one batch to create own gradient
+      Train();
+
+      // Put own gradient to peers queues
+      BroadcastGradients();
+
+      // Load own gradient
+      TensorVectorType gradients = g_.GetGradients();
+      TensorVectorType new_gradients;
+
+      // Sum all gradient in queue
+      while (!gradient_queue.empty())
+      {
+        {
```

Comment:
move this anonymous scope into a function call, e.g.:
void GetNewGradients(TensorVectorType & new_gradients)
{
          std::lock_guard<std::mutex> l(queue_mutex_);
          new_gradients = gradient_queue.front();
          gradient_queue.pop();
}